### PR TITLE
Displaying a specified asset

### PR DESCRIPTION
**Pull request by rowanwins**
This PR adds a new option called `displayAssetId` which when set looks for a COG and then displays it as a preference compared to some of the other assets. 
ATM it only works for COG assets but conceivably it could be extended to support other asset types.

### CODE REVIEW

**Comment by rowanwins** [comment]
PR includes a sample displaying the Usable Data Mask asset from a Planet item.

**Comment by rowanwins** [comment]
Related to #22

**Comment by m-mohr** [comment]
Just to get the scope right: This is meant to actually solve #22 by specifying an asset ID? Why the limitation to COGs and not the fully supported range of options in stac-layer?
I'm just wondering what the advantage/disadvantage is to specifying an ID instead of passing the whole asset object?

**Comment by rowanwins** [comment]
The asset definition alone doesn't contain the geometry, so if you want to show it's border you can't. This is also problematic for some asset types like the thumbnail or overview which don't have a geospatial reference without the broader item.

To support other asset types would be doable, I think it just needs a larger refactor to do so efficiently. Ideally I think there would be a method which took an asset definition and worked out the best way to render, whether COG or unreferenced image, or something else.

**Comment by m-mohr** [comment]
Yeah, but you could also just pass the full Item or Collection + the full assset, I assume. Then you don't need to fiddle around with IDs and finding them in the Item or Collection. It just makes the interface cleaner. You can just pass asset and/or item/collection objects and depending on what you pass, it makes the best out of it.

File: src/index.js [hunk 1]
```diff
@@ -459,6 +481,7 @@ const stacLayer = async (data, options = {}) => {
     // if we still haven't found a valid imagery layer yet, just add the first COG
     const cogs = Object.entries(assets).filter(entry => isAssetCOG(entry[1]));
     if (!addedImagery && cogs.length >= 1) {
+      console.log('adding a COG')
```
Comment (DanielJDufour) [inline src/index.js:484]: I assume this was accidentally left in, but if you want to log something here, I'd recommend mimicking the pattern used elsewhere with debug level and prefix in the message:
```js
if (debugLevel >= 1) console.log("[stac-layer] ...");
```

File: src/test/stac-item/cogs/cog-selected-asset.html [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+      // add OpenStreetMap basemap
+      L.tileLayer('http://{s}.tile.osm.org/{z}/{x}/{y}.png', {
+          attribution: '&copy; <a href="http://osm.org/copyright">OpenStreetMap</a> contributors'
+      }).addTo(map);
+
+
+      const url_to_stac_item = "https://raw.githubusercontent.com/cholmes/pdd-stac/beta2/disasters/hurricane-harvey/hurricane-harvey-0831/20170831_172754_101c/20170831_172754_101c.json";
+
+      fetch(url_to_stac_item).then(res => res.json()).then(async item => {
+        const lyr = await L.stacLayer(item, {
```
Comment (DanielJDufour) [inline src/test/stac-item/cogs/cog-selected-asset.html:33]: thanks for including a test!

File: src/index.js [hunk 3]
```diff
@@ -328,8 +331,27 @@ const stacLayer = async (data, options = {}) => {
...
+    if (displayAssetId !== null) {
+      const asset = assets[displayAssetId]
+      if (asset !== undefined && isAssetCOG(asset)) {
+        const href = toAbsoluteHref(asset.href);
+        try {
+          const georasterLayer = await createGeoRasterLayer(href, options);
+          if (debugLevel >= 1) console.log("[stac-layer] successfully created layer for", asset);
+          bindDataToClickEvent(georasterLayer, asset);
+          layerGroup.stac = { assets: [{ key: displayAssetId, asset }] };
+          setFallback(georasterLayer, () => addTileLayer({ asset, href, isCOG: true, isVisual: false, key: displayAssetId }));
```
Comment (DanielJDufour) [inline src/index.js:344]: nice use of fallback

**Review by DanielJDufour — APPROVED** [review]
Nice work.  I was concerned that it might add too much maintenance cost where as there are suitable workarounds.  However, I think you wrote this in a really clean way that reuses a lot of the functions used elsewhere.  I don't see this PR adding too much more complexity to the library.  If you find this functionality useful, then I say I approve (with the minor request to remove an excess log statement)

**Comment by DanielJDufour** [comment]
We could also leave this functionality undocumented ("private") for the next release, so you can use it a little more and have the flexibility of changing the option name or functionality if we decide to do so.

**Comment by m-mohr** [comment]
I'd like to better explain the architecture / interface that I had in mind:

Right now you have data and options as parameters for STAC Layer and data can be one of STAC Collection,     Item Collection,     STAC API Collections,     STAC API Items,     STAC Item,     STAC Asset or a list of STAC Assets.

These are all full STAC entities that have bbox or geometry except for the Assets and when you pass assets you usually want to show then in addition to the full STAC entity that is already shown.

So here's the idea: Instead of passing the assets as data, add a new parameter and remove them from the first parameter. So you have then 3 parameters:
1. `data`: STAC Collection, Item Collection, STAC API Collections, STAC API Items or STAC Item. You don't need to pass this and then if assets are present, you just show assets.
2. `assets`: A single asset or an array of assets (assets as objects, not IDs) that should be shown. By default this is empty/null and the automatic detection is used as implemented right now. If something is passed over, the assets are shown instead of the automatically added assets.
3. `options`

The clear advantage here is that you have a clean API that is very flexible and covers all cases (only assets, assets from multiple different sources (e.g. from multiple Items), only Item/Collection etc. Also, you don't have an issue if the ID can't be found because you always pass the full object, which makes the code simpler. Also, I think it is more future-proof for the changes you are planning to do in https://github.com/radiantearth/stac-browser/issues/12, @DanielJDufour, because you can use the array to pass over multiple assets already. I think my proposal doesn't need a lot of changes in the code, too.

Thoughts?

**Comment by DanielJDufour** [comment]
@m-mohr , I agree that your proposed change wouldn't require a lot of changes in the code.  To expand on it more, is this the API you are proposing?

```js
// data is a stac feature/item with multiple assets
// will default to using "visual" or first COG (not likely udm)
stacLayer(feature, undefined, { resolution: 128 });

// want to display a certain asset
stacLayer(feature, feature.assets.udm, { resolution: 128 });

// this also works as feature is ignored and bbox of udm cog is used
stacLayer(undefined, feature.assets.udm, { resolution: 128 });
```

**Comment by m-mohr** [comment]
@DanielJDufour Yes, exactly. For backward-compatibility, it might be better to have assets as third parameter though.

**Comment by DanielJDufour** [comment]
so something like this? (to be extra clear):
```js
// data is a stac feature/item with multiple assets
// will default to using "visual" or first COG (not likely udm)
stacLayer(feature, { resolution: 128 });

// want to display a certain asset
stacLayer(feature, { resolution: 128 }, feature.assets.udm);

// this also works as feature is ignored and bbox of udm cog is used
stacLayer(undefined,  { resolution: 128 }, feature.assets.udm);
```

**Comment by m-mohr** [comment]
Yes. The option to pass multiple assets as an array should be foreseen for the future: `stacLayer(feature, { resolution: 128 }, [feature.assets.udm, feature.assets.udm2]);`

**Comment by DanielJDufour** [comment]
Hi, I thought about it a little more and have the following thoughts:

I think we can handle both scenarios without too much trouble where a user wants to pass asset object(s) or the asset id(s).  If the users passes the asset ids we could then simply grab those from the feature. (I think it would only add one or a few lines of code).  That way we can leave it up to the user whatever makes the most sense to them.

@m-mohr for the example where you are passing in an array of assets, is this a scenario where each asset is for the same feature (i.e. same location and time) and you would like the first asset mapped to the red channel and the second asset mapped to the green channel?  That's doable with some configuration with GeoRasterLayer, but I just want to make sure I understand the use case.

I have a subjective personal preference for named arguments vs positional arguments, that's why I would be in favor of adding an `"assets"` or `"displayAssets"` optional parameter to the options object.  The syntax would thus be:

```js
// pass in asset(s)
// webVisual is an asset object { type: ..., href: ... } that should be used for visualizing on the web
stacLayer(feature, { assets: [webVisual] }) 

// bands are separated into different files like what Landsat 8 does
stacLayer(feature, { assets: [NIR, GREEN, BLUE] }) 

// pass in key
stacLayer(feature, { assets: ["webVisual"] }) 

// pass in keys
stacLayer(feature, { assets: ["nir", "green", "blue"] }) 

// display just an array of assets without reference to the parent feature
stacLayer([NIR, GREEN]) 
```

I've found that named arguments provide more flexibility in the long-term and wouldn't lock us in to holding the third position for assets.  Using named arguments (vs. changing positions) allows us the easily add new functionality without requiring major version upgrades.

On the naming of the variable, fwiw, I don't think we **need** to add "display" as in displayAssets because the display is assumed.  Titiler uses the same wording with their stac/tiles endpoint: https://developmentseed.org/titiler/endpoints/stac/#tiles just calling it "assets".  To the extent possible, I'd love to try to keep similarities with titiler and the other projects out there to improve interoperability.  And some day (my dream), have a standard syntax for expressing asset combinations and band math that works in Python and JavaScript. 

Open to your thoughts of course.  As you can tell, I borrowed heavily from both your ideas :-)

File: src/index.js (hunk 1 above)
Comment (rowanwins) [reply src/index.js]: removed :)

**Comment by rowanwins** [comment]
Have refactored this PR accordingly
- option name changed to `assets`
- accepts either a single thing or an array, and if it's supplied a single thing it converts it to an array
- accepts both asset keys and asset objects

**Comment by DanielJDufour** [comment]
I reflected on this a little more.  What's the use case for passing in an array of non-composite assets?  My understanding is that the primary use case for displaying multiple assets in one stacLayer is when the bands are separated into different files.  If we wanted to display multiple assets from different STAC features/items, then the implementer could just create separate stacLayer instances.  I do think there are some cool use cases of non-composite scenarios, like when a user wants to hide all the parts of an image that are "unsuable" based on a UDM, but I think those use cases would be better solved through a band math function (e.g. pixelValuesToColorFn for GeoRasterLayer).

GeoRasterLayer doesn't currently support automatic composite of bands separated into different files, but will shortly.  I have a working branch that I just need to run some final checks on before publishing.

fwiw, I would just accept one asset in the [asset] array at the moment, to satisfy your immediate need. And then we can upgrade to multiple assets once I get the new GeoRasterLayer version published.

**Comment by m-mohr** [comment]
I was thinking about passing a Collection + multiple Collection assets that could be shown side by side, but not sure whether that's actually a "common" use case... so I'd be fine with the proposal, I think.

**Comment by DanielJDufour** [comment]
No problem!  I hadn't fully thought out that use case.  I'm not familiar with collection assets.  I've only worked with feature assets.  Could you expand on what these assets do?  Is there an example you could share?

Is there a scenario where a user would want to visualize the collection-level assets along with the feature-level assets of the features of that collection?

**Comment by m-mohr** [comment]
Let's not overcomplicate things, many things can indeed also be solved with multiple stac layers, so let's keep it simple and go with the proposal above. I also don't really have real-world use-cases at hand.

**Review by DanielJDufour — APPROVED** [review]

**Comment by DanielJDufour** [comment]
Thanks for all your hard work, especially responding to my multiple rounds of feedback :-)

**Comment by DanielJDufour** [comment]
@rowanwins  and @m-mohr , new version is published to NPM!

**Comment by m-mohr** [comment]
Thank you!

**Pull request merged**